### PR TITLE
feat: Safety Filter support

### PR DESCRIPTION
**Pull request by bill-parker**
## Summary

This PR implements safety filter support for ADK, across three areas (General, Voice, Chat).
In Agent Studio, Content Safety Filters can be applied separately to
* General Configuration
* Voice Channel Settings
* Chat Channel Settings

This resource is implemented with a single shared base class `_BaseSafetyFilters` (itself extending YamlResource) that holds all the shared logic and three concrete subclasses:

- GeneralSafetyFilters
- ChannelSafetyFilters - class to handle filters derived from specific channels.
- VoiceSafetyFilters and ChatSafetyFilters - inherits from ChannelSafetyFilters, specifying their own channel-specific directories and projection paths.

Category data is handled by the _SafetyFilterCategory dataclass (with to_dict, from_dict, to_proto) and module-level helpers that translate between the three vocabularies: internal (enabled/precision), YAML/UI (enabled/level), and Azure projection (isActive/precision in camelCase keys).

ADK format uses `.yaml` to represent as thus:
```
enabled: true
categories:
  violence:
    enabled: true
    level: strict
  hate:
    enabled: true
    level: lenient
  sexual:
    enabled: true
    level: medium
  self_harm:
    enabled: false
    level: medium
```
NB: Channel settings have a global 'enabled' flag, while General settings do not.

## Motivation

Motivation is to improve parity with Agent Studio via UI.

## Changes
* Safety Filters Resource implemented including method to update protos.
* Sync client updated to read safety filters from projection data (General, Voice, Chat).
* Tests written to handle YAML round trip and other edge cases.

## Test strategy

Added unit tests, and conducted extensive testing with an AS project.

- [x] Added/updated unit tests
- [x] Manual CLI testing (`poly <command>`)
- [x] Tested against a live Agent Studio project
- [ ] N/A (docs, config, or trivial change)

## Checklist

- [x] `ruff check .` and `ruff format --check .` pass
- [x] `pytest` passes
- [x] No breaking changes to the `poly` CLI interface (or migration path documented)
- [ ] Commit messages follow [conventional commits](https://www.conventionalcommits.org/)

## Screenshots / Logs

QA - Validation test cases

incorrect category name:
Screenshot 2026-04-27 at 13 16 35
wrong values:
Screenshot 2026-04-27 at 13 14 07
missing field:
Screenshot 2026-04-27 at 13 15 14
not boolean for enabled:
Screenshot 2026-04-27 at 13 15 50
missing top level enabled:
Screenshot 2026-04-27 at 13 17 10

### CODE REVIEW

File: src/poly/resources/safety_filters.py [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+            if required not in category_data:
+                raise ValueError(
+                    f"Missing required field '{required}' for safety filter category "
+                    f"'{proj_key}' in azure config."
+                )
+        parsed[cat] = _SafetyFilterCategory(
+            enabled=category_data["isActive"],
+            precision=category_data["precision"],
+        )
+    return parsed
```
Comment (Ruari-Phipps) [inline src/poly/resources/safety_filters.py]: Maybe have this within the sync client, as it's only used there

File: src/poly/resources/safety_filters.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+        if required not in data:
+            raise ValueError(
+                f"Missing required field '{required}' for safety filter category '{cat_name}'."
+            )
+    level = data["level"]
+    matches = [k for k, v in PRECISION_MAPPING.items() if v == level]
+    if not matches:
+        valid_levels = ", ".join(sorted(PRECISION_MAPPING.values()))
+        raise ValueError(f"Invalid level '{level}'. Must be one of: {valid_levels}")
+    return {"enabled": data["enabled"], "precision": matches[0]}
```
Comment (Ruari-Phipps) [inline src/poly/resources/safety_filters.py]: Usually I have this sort of validation in `validate`
This will error on read

File: src/poly/resources/safety_filters.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+
+def _category_from_yaml_dict(data: dict, cat_name: str) -> dict:
+    """Translate a YAML/UI-vocab category dict to internal-vocab dict."""
+    for required in ("enabled", "level"):
+        if required not in data:
+            raise ValueError(
+                f"Missing required field '{required}' for safety filter category '{cat_name}'."
+            )
+    level = data["level"]
+    matches = [k for k, v in PRECISION_MAPPING.items() if v == level]
```
Comment (Ruari-Phipps) [inline src/poly/resources/safety_filters.py]: Could we just have a global inverse lookup dict

File: src/poly/resources/safety_filters.py [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+
+    @classmethod
+    def from_yaml_dict(
+        cls, yaml_dict: dict, resource_id: str, name: str, **kwargs
+    ) -> "_BaseSafetyFilters":
+        for required in ("enabled", "categories"):
+            if required not in yaml_dict:
+                raise ValueError(
+                    f"Missing required field '{required}' in safety filter config: {yaml_dict}."
+                )
```
Comment (Ruari-Phipps) [inline src/poly/resources/safety_filters.py]: Allow defaults

File: src/poly/resources/safety_filters.py (hunk 1 above)
Comment (bill-parker) [reply src/poly/resources/safety_filters.py]: shall I also remove the validate/error raising here (and replace with default vals if needed?) as we want to be more lenient on pull, than on push?

Comment (Ruari-Phipps) [reply src/poly/resources/safety_filters.py]: Yeah I think so

File: src/poly/resources/safety_filters.py (hunk 4 above)
Comment (bill-parker) [reply src/poly/resources/safety_filters.py]: I've moved this logic to validate().
If missing in the projection, it'll be replaced with defaults, but
I think it's still prudent to raise if it's missing from the local yaml.
as i.e. typos etc would not be caught properly

File: src/poly/docs/safety_filters.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
+# Safety Filters
```
Comment (AaronForinton) [inline src/poly/docs/safety_filters.md:1]: ty for adding this, lgtm

**Review by AaronForinton — APPROVED** [review]

File: src/poly/docs/docs.md [hunk 6]
```diff
@@ -15,6 +15,7 @@ Each project defines an AI voice or webchat agent. Resources in the project (flo
 │   ├── personality.yaml
 │   ├── role.yaml
 │   ├── rules.txt
+│   ├── safety_filters.yaml             # Optional
```
Comment (Ruari-Phipps) [inline src/poly/docs/docs.md]: Is this optional?

File: src/poly/resources/safety_filters.py [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+
+
+@dataclass
+class SafetyFilterCategory:
+    enabled: Optional[bool] = None
+    precision: Optional[str] = None
+
+    def to_dict(self) -> dict:
+        """Serialize using internal (backend) field names."""
+        return {"enabled": self.enabled, "precision": self.precision}
```
Comment (Ruari-Phipps) [inline src/poly/resources/safety_filters.py]: Does this duplicate `_category_to_yaml_dict` ?

File: src/poly/docs/docs.md (hunk 6 above)
Comment (bill-parker) [reply src/poly/docs/docs.md]: Ach, no it's not. It is optional for the others though,
can fix now

File: src/poly/resources/safety_filters.py (hunk 7 above)
Comment (bill-parker) [reply src/poly/resources/safety_filters.py]: I've removed that bit of dead code (to_dict) and refactored _category_to_yaml_dict, _category_from_yaml_dict as methods of SafetyFilterCategory.
This should be cleaner and more maintainable. Let me know what you think

**Review by Ruari-Phipps — APPROVED** [review]
Awesome work! LGTM!
